src/projection.jl

```diff
@@ -52,93 +128,197 @@ function ProjectTo(x::T) where {T<:Union{<:Tuple,NamedTuple}}
+    if axes(dy) == project.axes
+        return dy
+    else
+        for d in 1:max(N,M)
+            size(dy, d) == length(get(project.axes, d, 1)) || throw(DimensionMismatch("wrong shape!"))
+        end
+        return reshape(dy, project.axes)
+    end
 end
-function (project::ProjectTo{<:Array{T}})(dx::Tangent{<:SubArray}) where {T<:Number}
-    return project(dx.parent)
+# function (project::ProjectTo{<:Array{T}})(dx::Tangent{<:SubArray}) where {T<:Number}
+#     return project(dx.parent)
+# end
+
+# Row vectors -- need a bit more optimising!
+function ProjectTo(x::LinearAlgebra.AdjointAbsVec{T}) where {T<:Number}
+    sub = ProjectTo(parent(x))
+    ProjectTo{Adjoint{eltype(eltype(sub)), eltype(sub)}}(; parent=sub)
 end
+(project::ProjectTo{<:Adjoint})(dx::Adjoint) = adjoint(project.parent(parent(dx)))
+(project::ProjectTo{<:Adjoint})(dx::Transpose) = adjoint(conj(project.parent(parent(dx)))) # might copy twice?
+(project::ProjectTo{<:Adjoint})(dx::AbstractArray) = adjoint(conj(project.parent(vec(dx)))) # not happy!
+
+ProjectTo(x::LinearAlgebra.TransposeAbsVec{T}) where {T<:Number} = error("not yet defined")
 
 # Diagonal
-ProjectTo(x::T) where {T<:Diagonal} = ProjectTo{T}(; diag=ProjectTo(diag(x)))
-(project::ProjectTo{T})(dx::AbstractMatrix) where {T<:Diagonal} = T(project.diag(diag(dx)))
-(project::ProjectTo{T})(dx::AbstractZero) where {T<:Diagonal} = T(project.diag(dx))
+function ProjectTo(x::Diagonal)
+    eltype(x) == Bool && return ProjectTo(false)
+    sub = ProjectTo(diag(x))
+    ProjectTo{Diagonal{eltype(eltype(sub)), eltype(sub)}}(; diag=sub)
+end
+(project::ProjectTo{<:Diagonal})(dx::AbstractMatrix) = Diagonal(project.diag(diag(dx)))
+# (project::ProjectTo{T})(dx::AbstractZero) where {T<:Diagonal} = T(project.diag(dx))
 
-# :data, :uplo fields
-for SymHerm in (:Symmetric, :Hermitian)
+# Symmetric
+for (SymHerm, chk, fun) in ((:Symmetric, :issymmetric, :transpose), (:Hermitian, :ishermitian, :adjoint))
     @eval begin
-        function ProjectTo(x::T) where {T<:$SymHerm}
-            return ProjectTo{T}(; uplo=Symbol(x.uplo), parent=ProjectTo(parent(x)))
-        end
-        function (project::ProjectTo{<:$SymHerm})(dx::AbstractMatrix)
-            return $SymHerm(project.parent(dx), project.uplo)
-        end
-        function (project::ProjectTo{<:$SymHerm})(dx::AbstractZero)
-            return $SymHerm(project.parent(dx), project.uplo)
+        function ProjectTo(x::$SymHerm)
+            eltype(x) == Bool && return ProjectTo(false)
+            sub = ProjectTo(parent(x))
+            return ProjectTo{$SymHerm{eltype(eltype(sub)), eltype(sub)}}(; uplo=LinearAlgebra.sym_uplo(x.uplo), parent=sub)
         end
-        function (project::ProjectTo{<:$SymHerm})(dx::Tangent)
-            return $SymHerm(project.parent(dx.data), project.uplo)
+        function (project::ProjectTo{<:$SymHerm})(dx::AbstractArray)
+            dy = project.parent(dx)
+            dz = $chk(dy) ? dy : (dy .+ $fun(dy)) ./ 2
```

Comment:
:ok_hand:

Reply:
If we make it OK to mutate most gradients, then this can be done in-place and with N^2/2 not N^2 steps, sometimes. But later.

Reply:
If we really wanted
```suggestion
            dz = if $chk(dy)
                dy
            else
                k =  add!!(dy, $fun(dy))
                add!!(k, -1/2 .* k)
            end
```
Which will correctly handle the mutable/non-mutable array types.
We probably do not want.

Reply:
I don't think the in-place update can be done with broadcasting, it needs to iterate only one triangle so as not to step on its own toes.